codify-your-developer-vms.md

```diff
@@ -0,0 +1,83 @@
+---
+authorName: Torben Knerr
+authorGithubUsername: tknerr
+issue: 24
+title: Codify your Developer VMs!
+---
+
+# {{page.title}}
+
+Let’s talk about developer VMs, and why you absolutely and definitely must automate them (no excuses)! :)
+
+## What is a "Developer VM"?
+
+A developer VM is a virtual machine image that contains the complete development toolchain that is needed
+to work on a specific project (including compiler toolchains, IDEs, sytem wide settings, etc...).
+The aim is to ensure a consistent environment not only between team members of that project (which may have
+completely different operating systems on their laptops) but also with the build agents in the CI environment.
+
+## Why would I need a Developer VM anyway?
+
+Some people hate it, others love it. However you put it, it is one sure-fire way to ensure that the
+whole development team uses a consistent environment and avoids the typical “works on my machine”
+issues we all know about. Especially for us, executing and delivering a multitude of projects (and engineers
+joining and leaving these projects more or less frequently), developer VMs have become an indispensable tool.
+Not only for the consistency aspect, but also for reducing ramp-up times or being able to easily archive the
+development environment when the project enters the retirement phase.
+
+On regulated projects this is even more relevant, since you need to validate and verify the development toolchain
+and need to ensure that everyone uses exactly that defined toolchain.
+
+## OK, but why should I automate it?
+
+While we started out to craft these VMs manually (and with love) in the early days, we can nowadays easily
+automate the whole procedure with tools like Vagrant, Chef, Ansible and co. Applying automation and
+Infrastructure-as-Code principles to developer VMs not only increases transparency and reproducibility,
+but also allows for an in-VM update mechanism to maintain consistency while the toolchain evolves during
+the lifetime of the project.
+
+Finally, an automated developer VM should yield the following properties:
+
+* updatability - being able to update an existing VM with the push of a button from within the VM
+* testability - running automated tests in the VM to ensure that it works as expected
+* installability - the VM should distributed as a standard VirtualBox / VMware image
+* adaptability - developers should be able to easily adapt and improve the VM as the toolchain evolves
+* reproducibility - changes to the development environment are transparent, traceable and reproducible
+
```

Comment:
Gut und recht, aber der Versuch, die DevVM zu automatisieren ist in unserem Projekt grandios gescheitert. Es gab Probleme bei der MSSQL Installation (musste manuell ausgeführt werden weil nicht mit Chocolatey möglich), Office kann nicht installiert werden (weil die einen Teammembers eine Office 365 Lizenz haben, die anderen Office 2019 über MSDN), Windows an sich (schonmal eine Herausforderung mit der Lizenz) und dann und dann gibt es noch 1000 (oder sagen wir mal eine handvoll) Einstellungen welche doch noch manuell ausgeführt werden müssen: vs code plugins installieren, visual studio immer als admin ausführen, ...

Okay, genug Frust abgeladen. Den Artikel oben ist gut so und passt auch auf dem Artikel Level. Meine ganzen Umsetzungsfragen müssen ja da nicht beantwortet werden.

Reply:
Und fuer jeden der aufgelistet Probleme eine Liste. OK, ich habe gesagt ich werde Torbens Artikel reviewen, nicht die Kommentare dazu...